[Breaking] Refactor Ethereum token transfer signing input

catenocrypt opened the pull request:
## Description

BREAKING change.  Change in Eth signing input message, introducing specialized messages for different transfers:
native coin, ERC20, generic smart contract, etc.
Fixes #1176 .

## Breaking Change

**SigningInput for ETH transfer**:  The `amount` field is now moved to sub-message named `contractTransfer` of type `TransferContract`.  This is a breaking change!
Pay attention especially to JSON version, as the breaking change will not come out at compile time (signJSON()).

**SigningInput for ERC20 transfer**: This is a new addition; sub-message named `contractErc20` of type `ERC20TransferContract` should be used, with members `toAddress` and `amount`.  Building of the smart contract call is done by wallet-core, not the caller/app.

**SigningInput for smart contract call**: This is a new addition; for smart contract calls (except the special case of ERC20 above) a new sub-message named `contractGeneric` of type `GenericContract` should be used, with member `payload`.

See files for examples:

- src/proto/Ethereum.proto
- swift/Tests/Blockchains/EthereumTests.swift
- android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/ethereum/TestEthereumTransactionSigner.kt
- tests/Ethereum/TWAnySignerTests.cpp
- tests/Ethereum/SignerTests.cpp

## Testing instructions

Unit tests, esp. iOS&Android.

## Types of changes

<!--- What types of changes does your code introduce? Uncomment all the bullets that apply: -->

<!-- * Bug fix (non-breaking change which fixes an issue) -->

<!-- * New feature (non-breaking change which adds functionality) -->

* Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

<!--- The following points should be used to indicate the progress of your PR.  Put an `x` in all the boxes that apply right now, and come back over time and check them off as you make progress.  If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [ ] Update documentation as needed.
- [x] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain
- [x] If there is a related Issue, mention it in the description (e.g. Fixes #<issue_number> ).

hewigovens commented on src/proto/Ethereum.proto:
```diff
@@ -3,6 +3,32 @@ syntax = "proto3";
 package TW.Ethereum.Proto;
 option java_package = "wallet.core.jni.proto";
 
+message TransferContract {
```
I prefer to nest them under one message, see `Comos.proto`, this will reduce exposing too many proto types to swift/java

catenocrypt replied:
OK, done

hewigovens commented on src/Ethereum/Signer.cpp:
```diff
@@ -67,21 +71,72 @@ Signer::sign(const uint256_t &chainID, const PrivateKey &privateKey, const Data&
-        std::copy(address.bytes.begin(), address.bytes.end(), toAddress.data());
+    Data toAddress = addressStringToData(input.to_address());
+    switch (input.contract_oneof_case()) {
+        case TW::Ethereum::Proto::SigningInput::kContractTransfer:
```
```suggestion
        case Proto::SigningInput::kContractTransfer:
```

hewigovens commented on src/Ethereum/Signer.cpp:
```diff
@@ -67,21 +71,72 @@ Signer::sign(const uint256_t &chainID, const PrivateKey &privateKey, const Data&
+                return transaction;
+            }
+
+        case TW::Ethereum::Proto::SigningInput::ContractOneofCase::kContractGeneric:
```
```suggestion
        case Proto::SigningInput::ContractOneofCase::kContractGeneric:
```

hewigovens commented on src/Ethereum/Transaction.h:
```diff
@@ -26,8 +26,27 @@ class Transaction {
+        const Data& tokenContract, const Data& from, const Data& to, uint256_t tokenId);
+
+    // Create a generic smart contract transaction
+    static Transaction buildSmartContract(uint256_t nonce, uint256_t gasPrice, uint256_t gasLimit, const Data& to, const TW::Data& payload) {
```
```suggestion
    static Transaction buildSmartContract(uint256_t nonce, uint256_t gasPrice, uint256_t gasLimit, const Data& to, const Data& payload) {
```

hewigovens commented on src/proto/Ethereum.proto:
```diff
@@ -3,6 +3,32 @@ syntax = "proto3";
+}
+
+message ERC20TransferContract {
+    string to_address = 1;
```
just `to`
```suggestion
    string to = 1;
```

hewigovens commented on src/proto/Ethereum.proto:
```diff
@@ -3,6 +3,32 @@ syntax = "proto3";
+}
+
+message ERC721TransferContract {
+    string from_address = 1;
```
just `from` and `to`

hewigovens commented on src/proto/Ethereum.proto:
```diff
@@ -3,6 +3,32 @@ syntax = "proto3";
 
+message TransferContract {
+    // Amount to send in wei (256-bit number)
+    bytes amount = 1;
```
transfer could contain `data`

catenocrypt replied:
Really?  Then what's the difference between the geenric smart contract call an transfer?

catenocrypt replied:
OK, optional payload added

hewigovens commented on src/proto/Ethereum.proto:
```diff
@@ -3,6 +3,32 @@ syntax = "proto3";
 package TW.Ethereum.Proto;
 option java_package = "wallet.core.jni.proto";
 
+message TransferContract {
```
```suggestion
message Transfer {
```

hewigovens commented:
@catenocrypt @vikmeup since this is already a breaking change, let's use string for amount/gas limit/price?

vikmeup commented:
> @catenocrypt @vikmeup since this is already a breaking change, let's use string for amount/gas limit/price in `proto` level ?

As long as it's consistent across all other protos

catenocrypt commented:
> @catenocrypt @vikmeup since this is already a breaking change, let's use string for amount/gas limit/price in `proto` level ?

I would not change, only because of type change.

catenocrypt commented:
After some considerations, I reverted Aeternity and Aion changes, because they just create more clutter, and does not add more clarity.
Reverted commit:  https://github.com/trustwallet/wallet-core/pull/1210/commits/7d5a03b90e41965316475f09634591cd47268bb2

hewigovens commented on src/proto/Ethereum.proto:
```diff
@@ -3,6 +3,49 @@ syntax = "proto3";
 option java_package = "wallet.core.jni.proto";
 
+// Transaction payload (transfer, smart contract call, ...)
+message Payload {
```
```suggestion
message Transaction {
```

catenocrypt replied:
https://github.com/trustwallet/wallet-core/pull/1210/commits/3b87dcd014013c7aa5d4e279020c3015a55cc535

hewigovens commented on src/proto/Ethereum.proto:
```diff
@@ -3,6 +3,49 @@ syntax = "proto3";
+        bytes amount = 1;
+
+        // Optional payload
+        bytes payload = 2;
```
```suggestion
        bytes data = 2;
```

hewigovens commented on src/proto/Ethereum.proto:
```diff
@@ -3,6 +3,49 @@ syntax = "proto3";
+    // Generic smart contract transaction
+    message ContractGeneric {
+        // Contract call payload
+        bytes payload = 1;
```
```suggestion
        bytes data = 1;
```

hewigovens commented on src/proto/Ethereum.proto:
```diff
@@ -3,6 +3,49 @@ syntax = "proto3";
+
+    oneof payload_oneof {
+        Transfer payload_transfer = 1;
+        ERC20Transfer payload_erc20 = 2;
```
can you also add `ERC20Approve`?

catenocrypt replied:
Added

hewigovens commented on src/proto/Ethereum.proto:
```diff
@@ -3,6 +3,58 @@ syntax = "proto3";
+    }
+
+    oneof transaction_oneof {
+        Transfer transaction_transfer = 1;
```
```suggestion
        Transfer transfer = 1;
```

no need to add prefix `transaction_`? , looks verbose

catenocrypt replied:
I've copied the "oneof ENTITY_oneof { type1 ENTITY_type1 = 1; type2 ENTITY_type2 = 2; ... }" pattern seen in some other .proto file.

catenocrypt replied:
Removed

hewigovens commented:
@catenocrypt @vikmeup If we merge this, we need to update TrustSDK + iOS app at the same time, and ask 3rd party apps to update SDK as well
